iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,325 @@
+use std::io::{Error, ErrorKind};
+
+use async_stream::stream;
+use futures::Stream;
+use rand::{Rng, RngCore};
+use tokio::io;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
+use tokio::net::TcpStream;
+use ursa::kex::KeyExchangeScheme;
+use ursa::kex::x25519::X25519Sha256;
+use ursa::keys::{PrivateKey, PublicKey};
+
+use iroha_actor::{Actor, Addr, Context, Handler};
+use iroha_data_model::prelude::PeerId;
+use iroha_logger::warn;
+
+use crate::message::{Message, MessageResult};
+use crate::network::{ReceivedPeerMsg, SendPeerMsg};
+use crate::Network;
+use ursa::encryption::symm::{SymmetricEncryptor, Encryptor};
+
+const MAX_MESSAGE_LENGTH: usize = 2 * 1024 * 1024;
+const MAX_HANDSHAKE_LENGTH: usize = 256;
+pub const DEFAULT_AAD: &[u8; 12] = b"Iroha2Iroha2";
+
+pub struct Peer<E: Encryptor + Send + 'static> {
+    id: PeerId,
+    read: Option<OwnedReadHalf>,
+    write: Option<OwnedWriteHalf>,
+    state: State,
+    secret_key: PrivateKey,
+    public_key: PublicKey,
+    cipher: Option<SymmetricEncryptor<E>>,
+    network_addr: Addr<Network<E>>,
+}
+
+impl<E: Encryptor + Send + 'static> Peer<E> {
+    pub fn new(id: PeerId, stream: Option<TcpStream>, state: State, addr: Addr<Network<E>>) -> Self {
+        // P2P encryption primitives
+        let x25519 = X25519Sha256::new();
+        let (public_key, secret_key) = x25519.keypair(None).expect("Unexpected error generating keypair");
```

Comment:
Could you call here only encryptor from generic? Also it is bad to panic

Reply:
Second generic? I thought that we agreed on this algo, and will not change it.
Also, this is not expected to fail ever.

Reply:
I mean not only encryptor, but [`KeyExchangeScheme`](https://docs.rs/ursa/0.3.6/ursa/kex/trait.KeyExchangeScheme.html) might change in future.

Reply:
Checked out sorces of keypair:
https://docs.rs/ursa/0.3.6/src/ursa/kex/x25519.rs.html#18-49

Seems that they don't throw right now. But they might change code and do that in future. Peer should handle this case without panicking

Reply:
This functions always produce results. But is there is an error, what kind of recovery we can think of?

Reply:
This is constructor of peer. You can push errors to network itself.